apollo-federation/src/sources/connect/validation/entity/keys.rs

```diff
@@ -0,0 +1,215 @@
+                    .values()
+                    .flatten()
+                    .map(|fs| {
+                        format!(
+                            "... on {} {}",
+                            fs.selection_set.ty,
+                            fs.selection_set.serialize().no_indent()
+                        )
+                    })
+                    .collect_vec(),
+            )
+            .finish()
+    }
+}
+
+/// Given the parts of the connector that contain variables, synthesize a field
+/// set appropriate for use in a @key directive.
+///
+/// This only looks at `$args` variables because those are the ones used
+/// for `entity: true` connectors.
+///
+/// TODO: this is simpler to code in expand/mod.rs and might be something we
+/// could combine. The expand code not only generates a key, but also copies
+/// the types related to the key into the expanded subgraph, so it does more work
+/// than this function.
+pub(crate) fn make_key_field_set_from_variables(
+    schema: &Schema,
+    object_type_name: &Name,
+    variables: &[VariableReference<Namespace>],
+    resolver: EntityResolver,
+) -> Result<Option<Valid<FieldSet>>, Message> {
+    let params = variables
+        .iter()
+        .filter(|var| match resolver {
+            EntityResolver::Explicit => var.namespace.namespace == Namespace::Args,
+            EntityResolver::Implicit => var.namespace.namespace == Namespace::This,
+        })
+        .unique()
+        .collect_vec();
+
+    if params.is_empty() {
+        return Ok(None);
+    }
+
+    let mut keys = Vec::with_capacity(params.len());
+    for VariableReference { path, .. } in params {
+        let parts = path.iter().map(|part| part.part.as_ref());
+        let field_and_selection = FieldAndSelection::from_path(parts);
+        keys.push(field_and_selection.to_key());
+    }
+
+    FieldSet::parse_and_validate(
+        Valid::assume_valid_ref(schema),
+        object_type_name.clone(),
+        keys.join(" "),
+        "",
+    )
+    // This shouldn't happen because we've already validated the inputs using ArgumentVisitor
+    .map_err(|_| Message {
+        code: Code::GraphQLError,
+        message: format!("Variables used in connector (`{}`) for `{}` cannot be used to create a valid `@key` directive.", variables.iter().join("`, `"), object_type_name),
+        locations: vec![],
+    }).map(Some)
+}
+
+/// Represents a field and the subselection of that field, which can then be joined together into
+/// a full named selection (which is the same as a key, in simple cases).
+///
+/// TODO: this is copied from expand/mod.rs
+struct FieldAndSelection<'a> {
+    field_name: &'a str,
+    sub_selection: String,
+}
+
+impl<'a> FieldAndSelection<'a> {
+    /// Extract from a path like `a.b.c` into `a` and `b { c }`
+    fn from_path<I: IntoIterator<Item = &'a str>>(parts: I) -> Self {
+        let mut parts = parts.into_iter().peekable();
+        let field_name = parts.next().unwrap_or_default();
```

Comment:
This seems like silent failure if the path was empty - it will end up with a `FieldAndSelection` whose `field_name` is the empty string.

Reply:
i have a plan to get rid of this code entirely